src/open_inwoner/js/components/siteimprove/error-tracking.js

```diff
@@ -0,0 +1,133 @@
+/**
+ * Note: users can upload multiple faulty files with their respective errors, and can delete them one by one,
+ * Tracking should only happen on newly added errors, not the entire occurring set.
+ */
+// if (typeof _sz === 'undefined') {
+//   /** Mock SiteImprove `_sz` object for testing - only used during development */
+//   var _sz = {
+//     push: function (data) {
+//       try {
+//         console.log('Event pushed to _sz:', data)
+//       } catch (error) {
+//         console.error('Error occurred while pushing event data:', error)
+//       }
+//     },
+//   }
+// }
+
+/**
+ * Class that handles the transaction between file errors and SiteImprove.
+ * Used in `/mijn-aanvragen/{number}/{uuid}/status`
+ */
+class DynamicFileInputErrors {
+  constructor() {
+    this.initialized = false
+    this.previousErrorState = new Map()
+    this.bindEvents()
+  }
+
+  /**
+   * Binds events to callbacks.
+   * Use this to define EventListeners, MutationObservers etc.
+   */
+  bindEvents() {
+    if (this.#formElement && !this.initialized) {
+      this.initialized = true
+      this.#formElement.addEventListener(
+        'change',
+        this.handleChanges.bind(this)
+      )
+    }
+  }
+
+  /**
+   * Gets called when this.getForm() changes.
+   * @param {MouseEvent} event
+   */
+  handleChanges(event) {
+    if (!this.#fileInputElement || event.target !== this.#fileInputElement)
+      return
+
+    const currentErrors = Object.entries(this.#occurringErrors).reduce(
+      (acc, [key, nodes]) => {
+        nodes.forEach((node) => {
+          const fileName =
+            node.querySelector('.file__name').textContent ??
+            [...this.#fileListElement.children].indexOf(node)
```

Comment:
I am not sure I follow the logic behind the fallback value: this would simply return an empty ID no? Perhaps something like `unnamed-file-{index}` would be better? I suppose most browsers won't let you select unnamed files, so this is really just a guard, but still.

Also: isn't the spread operator redundant here?

```suggestion
            node.querySelector('.file__name').textContent ??
            this.#fileListElement.children.indexOf(node)
```

Reply:
This could better be answered by @stefrado but here is my take:
in almost all cases, the .file__name element should contain some text in reality. However, the fallback (indexOf(node)) is a safeguard for unexpected edge cases:
-If there's a rendering delay and the .file__name text isn't populated yet.
-If the .file__name selector is somehow missing due to a DOM update issue (like flakey HTMX).
-If the system processes files asynchronously and some file metadata is not yet available.
While unlikely, having a fallback (here the fallback is an index number) ensures that error tracking doesn't break due to an empty identifier.
Just the ID is indescriptive so yeah, unnamed-file-{index} could be better but humans won't actually see that, it's code only.

The spread operator isn't quite redundant. Since "this.#fileListElement.children" is an **HTMLCollection**, but the .indexOf() method only works on arrays, so the code is converting the collection into an array before calling .indexOf(node). The spread operator creates a shallow copy.
A more readable way to do this, would be `const index = Array.from(this.#fileListElement.children).indexOf(node)` but you'll still need it.

All together it could be something like this:
`const currentErrors = Object.entries(this.#occurringErrors).reduce(
  (acc, [key, nodes]) => {
    nodes.forEach((node) => {
      const index = Array.from(this.#fileListElement.children).indexOf(node)
      const fileName = node.querySelector('.file__name')?.textContent || `unnamed-file-${index}`
      acc.set(fileName, this.#ERROR_MAP[key])
    })
    return acc
  },
  new Map()
)
`

Reply:
That's a very useful explanation, thanks. I take your point that the fallback is really more of a theoretical option, so feel free to ignore.

I didn't realize that `HTMLCollection` didn't implement the full array interface, good to know :+1: .

Reply:
That may be one of the reasons why React code is chockfull of spread operators.